README
README for Clutter 1.26.2
===============================================================================

Clutter is an open source software library for creating fast, compelling,
portable, and dynamic graphical user interfaces.

REQUIREMENTS
-------------------------------------------------------------------------------

Clutter currently requires:

  • GLib ≥ 2.44.0
  • Cogl ≥ 1.21.2
  • JSON-GLib ≥ 0.12.0
  • Atk ≥ 2.5.3
  • Cairo ≥ 1.14.0
  • PangoCairo ≥ 1.30

When building the X11 backend, Clutter depends on the following extensions:

  • XComposite ≥ 0.4
  • XDamage
  • XExt
  • XInput (1.x or 2.x)
  • XKB

When building the Wayland backend, Clutter also depends on:

  • wayland-client
  • xkbcommon

When building the GDK backend, Clutter also depends on:

  • gdk-3.0 > 3.22.6

When building the CEx100 backend, Clutter also depends on:

  • libgdl

When building the evdev input backend, Clutter also depends on:

  • xkbcommon
  • libudev ≥ 136
  • libinput ≥ 0.19.0

If you are building the API reference you will also need:

  • GTK-Doc ≥ 1.20

If you are building the additional documentation you will also need:

  • xsltproc
  • jw (optional, for generating PDFs)

If you are building the Introspection data you will also need:

  • GObject-Introspection ≥ 1.39.0

GObject-Introspection is available from:

  git://git.gnome.org/gobject-introspection

RESOURCES
-------------------------------------------------------------------------------

The official Clutter website is:

   http://www.clutter-project.org/

The API references for the latest stable release are available at:

   https://developer.gnome.org/clutter/stable/
   https://developer.gnome.org/cogl/stable/

The Clutter Cookbook is available at:

   https://developer.gnome.org/clutter-cookbook/

New releases of Clutter are available at:

   https://download.gnome.org/sources/clutter/

The Clutter blog is available at:

   http://www.clutter-project.org/

To subscribe to the Clutter mailing list and read the archives, use the
Mailman web interface available at:

   https://mail.gnome.org/mailman/listinfo/clutter-list

New bug page on Bugzilla:

   https://bugzilla.gnome.org/enter_bug.cgi?product=clutter

Clutter is licensed under the terms of the GNU Lesser General Public
License, version 2.1 or (at your option) later.

BUILDING AND INSTALLATION
-------------------------------------------------------------------------------

To build Clutter from a release tarball, the usual autotool triad should
be followed:

  $ ./configure
  $ make
  # make install

To build Clutter from a Git clone, run the autogen.sh script instead
of the configure one. The autogen.sh script will run the configure script
for you, unless the NOCONFIGURE environment variable is set to a non-empty
value.

See also the wiki page:

  https://wiki.gnome.org/Projects/Clutter

Clutter has additional command line options for the configure script:

 --enable-debug=[no/minimum/yes]
	Controls Clutter debugging level:

	yes:
                All GLib asserts, checks and support for runtime Clutter
                debugging notes through CLUTTER_DEBUG. This is the default
                value for developers snapshots.

        minimum:
                Just GType cast checks and support for runtime Clutter
                debugging notes through CLUTTER_DEBUG. This is the default
                for stable releases.

        no:
                No GLib asserts or checks and no support for runtime Clutter
                debugging notes. Only use in extreme performance and/or size
                optimization cases, though it is strongly discouraged.

 --enable-maintainer-flags=[no/yes/error]
        Use strict compiler flags. This defaults to 'yes' for building from
        Git to 'no' for tarball releases. If 'error' is used, then -Werror
        will be enabled (if available).

 --enable-gtk-doc
	use gtk-doc to build API documentation (default=no). Requires gtk-doc
	present on the target system.

 --enable-docs=[no/yes]
        Build additional documentation. Requires xsltproc for DocBook
        conversion, and optionally jw for PDF generation.

 --enable-gcov=[no/yes]
        Build Clutter with coverage report support, provided by gcov. This
        feature only works with the GNU Compiler Suite and gcov installed.

 --disable-tests
        Disable building the Clutter tests suite.

 --disable-examples
        Disable building the Clutter API reference examples.

 --enable-deprecated=[yes/no]
        Whether deprecated symbols should be available when compiling Clutter.

 --disable-Bsymbolic
        Disable linking with -Bsymbolic.

 --enable-x11-backend=[yes/no/check]
        Enable the X11 backend. (default=check)

 --enable-win32-backend=[yes/no/check]
        Enable the native Microsoft Windows backend. (default=check)

 --enable-quartz-backend=[yes/no/check]
        Enable the native Quartz backend. (default=check)

 --enable-gdk-backend=[yes/no/check]
        Enable the GDK backend. (default=check)

 --enable-wayland-backend=[yes/no]
        Enable the Wayland client backend. (default=no) [EXPERIMENTAL]

 --enable-cex100-backend=[yes/no]
        Enable the CEx100 platform backend. (default=no) [EXPERIMENTAL]

 --enable-egl-backend=[yes/no]
        Enable the EGL framebuffer backend. (default=no)

 --enable-tslib-input=[yes/no]
        Enable the TSLib input backend. (default=no) [EXPERIMENTAL]

 --enable-evdev-input=[yes/no]
        Enable the evdev input backend. (default=no) [EXPERIMENTAL]

See also the INSTALL file generated by autotools for further information.

VERSIONING
-------------------------------------------------------------------------------

Clutter uses the common "Linux kernel" versioning system, where
even-numbered minor versions are stable and odd-numbered minor
versions are development snapshots.

Different major versions break both API and ABI but are parallel
installable. The same major version with differing minor version is
expected to be ABI compatible with other minor versions; differing
micro versions are meant just for bug fixing. On odd minor versions
the newly added API might still change.

The micro version indicates the origin of the release: even micro
numbers are only used for released archives; odd micro numbers are
only used on the Git repository.

HACKING
-------------------------------------------------------------------------------

If you want to hack on and improve Clutter check the HACKING file for
general implementation guidelines, and the HACKING.backends for
backend-specific implementation issues.

The CODING_STYLE file contains the rules for writing code conformant to the
style guidelines used throughout Clutter. Remember: the coding style is
mandatory; patches not conforming to it will be rejected by default.

BUGS
-------------------------------------------------------------------------------

Bugs should be reported to the Clutter Bugzilla at:

  http://bugzilla.gnome.org/enter_bug.cgi?product=clutter

You will need a Bugzilla account.

In the report you should include:

  • what system you're running Clutter on;
  • which version of Clutter you are using;
  • which version of GLib, Cogl, and OpenGL (or OpenGL ES) you are using;
  • which video card and which drivers you are using, including output of
    glxinfo and xdpyinfo (if applicable);
  • how to reproduce the bug.

If you cannot reproduce the bug with one of the tests that come with Clutter
source code, you should include a small test case displaying the bad
behaviour.

If the bug exposes a crash, the exact text printed out and a stack trace
obtained using gdb are greatly appreciated.

CONTRIBUTING
-------------------------------------------------------------------------------

Patches should be submitted using Bugzilla. Patches fixing a bug should be
attached to the bug report; patches for new features or for fixing bugs not
yet reported should be attached to a newly opened bug.

Patches should always be in the unified diff format, using:

  diff -Nuarp clutter.source clutter.patched > clutter-patch.diff

If diffing against the Git repository, you should use:

  git diff > clutter-patch.diff

Or, better: commit locally and use `git format-patch` to generate a patch
containing authorship details, so that members of the Clutter development
team can credit your contribution properly.

Another useful tool for interacting with Git and Bugzilla is git-bz(1):

  http://git.fishsoup.net/man/git-bz.html

Which is available here:

  http://git.fishsoup.net/cgit/git-bz/

Patches submitted against Clutter have to pass the conformance test suite, or,
if possible, add new unit tests for the conformance test suite in case of new
features. Ensure you run the conformance test suite every for every patch you
wish to submit, by using:

  make -C tests/conform check

and verifying that the whole test suite passes.

RELEASE NOTES
-------------------------------------------------------------------------------

Relevant information for developers with existing Clutter applications
wanting to port to newer releases (see NEWS for general information on new
features).

Release Notes for Clutter 1.26
-------------------------------------------------------------------------------

• The evdev backend specific ClutterPointerConstrainCallback type was changed
  to include not only the target position but also the previous position. This
  is an API change, and will require a version check in any caller code.

Release Notes for Clutter 1.24
-------------------------------------------------------------------------------

• ClutterModel, ClutterModelIter, and ClutterListStore have been deprecated.
  You're strongly encouraged to use the GListModel interface, and the
  GListStore class, as a replacement.

• The macros used to access the state flags in ClutterActor have been
  deprecated, and replaced by equivalent functions.

• Clutter types now can be used with the `g_autoptr()` and `g_auto()`
  macros provided by GLib and supported by GCC and Clang.

• The old, non-namespaced CLUTTER_* key symbols now have an associated
  deprecation warning.

• The private, unused ClutterX11XInputEventTypes enumeration, which was
  erroneously placed in the clutter-x11.h public header, has been removed.

• The Wayland client backend is now autodetected depending on whether the
  required dependencies are available.

Release Notes for Clutter 1.22
-------------------------------------------------------------------------------

• The default backend for Clutter starting from this release is the GDK
  backend, if Clutter was compiled with it. This allows better integration
  with the underlying windowing system platform on X11 and Wayland. On
  Windows and MacOS the native backends are still the preferred ones. If you
  require a specific backend you should use clutter_set_windowing_backend()
  with the CLUTTER_WINDOWING_* backend you require, prior to call
  clutter_init().

• The ClutterRequestMode enumeration has a new value. Code checking for the
  values of ClutterRequestMode should already be resilient to changes to this
  enumerations, like for all enumerations in Clutter.

Release Notes for Clutter 1.20
-------------------------------------------------------------------------------

• The clutter_stage_set_paint_callback() experimental function has been
  removed from the Clutter ABI; it has been replaced by the ::after-paint
  signal on the ClutterStage class. The set_paint_callback() method was
  marked as "experimental", required a special definition to be usable, and
  no guarantees were made on its continued existence.

Release Notes for Clutter 1.18
-------------------------------------------------------------------------------

• Until 1.18, ClutterStage removed its children during its dispose()
  implementation, before the default ClutterActor::destroy() implementation
  would run. ClutterStage will now destroy the children when it is destroyed
  to ensure that the children are destroyed, and that custom code can remove
  references through the ClutterActor::destroy signal.

• Clutter does not depend on the XFIXES extension API on X11 any more. Before
  1.18 Clutter used the XFIXES API to hide the cursor; the API is less than
  useful for toolkits and applications, so Clutter unconditionally uses the
  fall back code that was in place in case XFIXES was not available.

• ClutterText emits the ::insert-text and ::delete-text signals before the
  contents of the ClutterTextBuffer are changed, as documented. The signal
  emission cannot be guaranteed if the ClutterTextBuffer API is used instead
  of the ClutterText API.

• Starting from 1.18, the Clutter evdev input device backend no longer uses
  libevdev and libgudev directly, but relies on libinput for discovering,
  reading and processing input devices.

• The Clutter evdev input device backend was already considered
  experimental and not subject to Clutter's API and ABI stabitility
  guarantees. Starting from 1.18, users have to explicitly acknowldge
  this by having to #define CLUTTER_ENABLE_COMPOSITOR_API to use its
  public API.

Release Notes for Clutter 1.16
-------------------------------------------------------------------------------

• Implicit transitions will not be created on actors that are not mapped,
  unless they are in a cloned branch of the scene graph. This was never an
  approved case, as ClutterActor would implicitly skip layout and paint on
  unmapped actors, but now it's being enforced through the animation machinery
  as well. Using explicit transitions still works, as explicit transitions
  completely place the developer in charge.

Release Notes for Clutter 1.14
-------------------------------------------------------------------------------

• The X11 backend (both core X events and XInput 2 backends) now mask out
  scroll lock and num lock modifiers from the event state.

Release Notes for Clutter 1.12
-------------------------------------------------------------------------------

• ClutterBinLayout no longer requests or allocates size for children that are
  not visible, which is how the other layout managers work.

• The expansion and alignment layout properties in all the layout managers
  that provide them have been deprecated in favour of the equivalent ones
  on ClutterActor; the ClutterLayoutManager implementations provided by
  Clutter have been updated to honour the ClutterActor:x-expand,
  ClutterActor:y-expand, ClutterActor:x-align, and ClutterActor:y-align
  properties, if set.

• Setting the ClutterActor:fixed-position-set to FALSE will reset any
  currently set fixed position to (0, 0).

• ClutterActor's margin properties are now animatable.

• ClutterBindingActionFunc typedef now has a user_data parameter; the user
  data was passed by every caller, and ClutterBindingActionFunc was not used
  anywhere in the API except for documentation purposes.

• The clutter_threads_enter() and clutter_threads_leave() functions have been
  deprecated; the lead to non portable code, and encourage broken behaviour
  with regards to threaded applications. The only supported, portable way of
  writing Clutter application employing threads is to defer long running
  operations to a worker thread, and schedule UI updates to the main loop at
  well defined synchronization points.

Release Notes for Clutter 1.10
-------------------------------------------------------------------------------

• The ClutterActor::paint, ClutterActor::queue-redraw, and
  ClutterActor::queue-relayout signals are now annotated as not allowing
  signal emission hooks; this enables some optimization inside GLib, and
  given the amount of emissions these three signals have, we want to get
  through all the fast paths we can.

• The ClutterActorBox parameter of the ClutterActor::allocation-changed
  signal is now marked as G_SIGNAL_TYPE_STATIC_SCOPE.

• The clutter-stage-window.h header is not installed any more; this header
  was never meant to be public in the first place, and nothing could have
  been implementing the ClutterStageWindow interface and use it with Clutter.

• ClutterActor will call its unmap() implementation when it is removed from
  its parent. This may happen after the ClutterActor::destroy signal has been
  emitted, i.e. during its dispose(). Prior to 1.10, calling the
  clutter_actor_destroy() function would unmap the actor first, and then
  emit the destroy signal. This means that extra care should be used when
  overriding the map() and unmap() virtual functions, to avoid dereferencing
  NULL pointers. It is also worthy of note that since Clutter 1.8, overriding
  map() or unmap() is not required any more for subclasses of ClutterActor
  that have children, as ClutterActor will automatically do the right thing
  inside its own implementation, and map or unmap its children when needed.

• ClutterBox and ClutterGroup have been deprecated. ClutterActor should be
  used directly, instead. Since ClutterStage inherits from ClutterGroup, the
  instance and class structures are still available, but subclassing Group
  is strongly disencouraged.

• ClutterContainer provides default implementations for add(), remove(),
  foreach(), raise(), lower(), and sort_depth_order(); this means that
  overriding these virtual functions is now deprecated, and developers
  should only implement ClutterContainer for classes providing child
  properties.

• ClutterActor is not an abstract type any more, and can be instantiated
  either through clutter_actor_new() or through g_object_new().

• ClutterActor now implements ClutterContainer, and takes over the entire API
  for modifying the scene graph; this means that every actor can have children
  and it's not necessary any more to implement the ClutterContainer interface.
  ClutterActor provides new API to replace clutter_actor_set_parent() and
  clutter_actor_unparent(), as well as a default implementation of every
  Container virtual function. Existing actors overriding ClutterContainer and
  providing an alternate implementation will continue to work, even though it
  is strongly encouraged to port to the new API.

• The ClutterActor::destroy signal is going to be emitted at the beginning
  of the dispose sequence of a ClutterActor, instead of the end. This allows
  to access the state of the actor being destroyed, instead of just being
  able to use its pointer. This may expose bugs in code that does not check
  for NULL actor pointers.

• The long since broken depth cueing support in ClutterStage has been
  deprecated; the fixed pipeline fog API has been pretty much useless
  since the switch to colors and textures with premultiplied alpha
  channels. Setting ClutterStage:use-fog to TRUE has no visible results,
  and querying the :use-fog and :fog properties will always yield the
  default values.

• ClutterDeformEffect switched from using CoglVertexBuffer to using the
  CoglPrimitive API internally, to improve performance and use non-deprecated
  Cogl API. CoglPrimitive converts COGL_WRAP_MODE_AUTOMATIC to
  COGL_WRAP_MODE_CLAMP_TO_EDGE, unlike CoglVertexBuffer which converts it to
  COGL_WRAP_MODE_REPEAT. This prevents artifacts when sampling texture
  coordinates outside the [ 0, 1 ] range. This change may cause the back
  texture to not be painted if its coordinates go outside the allowed range,
  for instance when using a custom transformation matrix on the back material
  used by the ClutterDeformEffect.

• The "default stage" has been deprecated; since the 1.0 release, the default
  stage creation was deferred to the call to clutter_stage_get_default(), and
  the preferred way for getting a ClutterStage was calling clutter_stage_new()
  instead. On platforms that do not support multiple stages, attempting to
  create more than one stage will lead to an error, and Clutter will abort.

• Clutter can support multiple backends in the same shared library. Only one
  windowing or input backend can be used at run time. As a result of this
  change, the shared library name used by Clutter has changed from:

        libclutter-<flavour>-<API version>.so

  to:

        libclutter-<API version>.so

  The pkg-config file has been updated accordingly. Until the next major API
  break, Clutter will ship compatibility links for all the previous "flavours"
  that were available in versions < 1.10; this allows applications dynamically
  linking against Clutter, or using dlopen(), to keep working. For libraries

NEWS
Clutter 1.26.2                                                       2017-05-12
===============================================================================

  • List of changes since Clutter 1.26.0

    - Accept upper case 'A' when selecting all content in a Text actor
    This is similar to what GTK+ allows, and makes it easier to select text
    with Caps Lock enabled.

    - Use GDK API when using Wayland sub-surfaces
    The GDK backend on Wayland uses sub-surfaces to allow Cogl to manage the
    rendering surface; now that GDK exposes the necessary API, we can avoid
    using the Wayland API directly, and instead let GDK manage the surface
    for us.

    - Documentation fixes

    - Translation updates
    Danish, Basque, Chinese (Traditional), Occitan, Indonesian, Polish,
    Spanish, Slovak, Kazakh, British English, Croatian, Hungarian, Norwegian
    bokmål, Belarusian.

  • List of bugs fixed since Clutter 1.26.0

    #766326 - text: Bind <ctrl>A in addition to <ctrl>a
    #768243 - Don't create the Cogl GLib source multiple times
    #768734 - actor: Clarify that add_effect() sinks floating references
              on the effect
    #781975 - Under Wayland, Totem spawns a new window without controls for
              some files and then gnome-shell crashes
    #759085 - seg fault in actor-graph test
    #774546 - [Wayland][gdk][stage] Use GDK API instead of Wayland directly
    #762888 - [PATCH] backend: check for wayland before x11
    #772412 - clutter-script: Document support for top-level arrays
    #770288 - clutter-script: Add an example of translation in ClutterScript
    #768898 - clutter: Fix typo in docs for ClutterActor::queue-relayout

Many thanks to:

  Piotr Drąg, Philip Withnall, Olivier Fourdan, Cédric Valmary, Fabio Tomat,
  Andika Triwidada, Ask Hjorth Larsen, Baurzhan Muftakhidinov, David King,
  Dušan Kazik, Florian Müllner, Gustavo Noronha Silva, Gábor Kelemen,
  Inaki Larranaga Murgoitio, Jeff Bai, Jeremy Bicha, Kjartan Maraas,
  Marc-Antoine Perennou, Nigel Taylor, Owen W. Taylor, Peter Mráz,
  Yuras Shumovich, gogo


Clutter 1.26.0                                                       2016-03-21
===============================================================================

  • List of changes since Clutter 1.25.6

    - Avoid spamming accessibility state changes in clones
    Clones of unmapped actors will cause a temporary mapping of their source,
    which will result in spurious accessibility events.

    - Ignore GDK deprecation compiler warnings

    - Resync build configuration for Windows

    - Translation updates
    Serbian, Swedish, Catalan, Occitan, Greek, Latvian, Russian, Danish

  • List of bugs fixed since Clutter 1.25.6

    #756371 - Accessible event spam in Overview

Many thanks to:

  Jordi Mas, Anders Jonsson, Ask Hjorth Larsen, Chun-wei Fan, Cédric Valmary,
  Efstathios Iosifidis, Rūdolfs Mazurs, Stas Solovey, Марко Костић,
  Мирослав Николић

Clutter 1.25.6                                                       2016-02-18
===============================================================================

  • List of changes since Clutter 1.25.4

    - Enable swap throttling for fullscreen windows with GDK
    Avoids performance penalties with the GDK backend on constrained devices
    running full screen Clutter and Clutter-GTK applications.

    - Add platform-specific data for the libinput input backend
    This allows implementing high level functionality when writing Wayland
    compositors based on Clutter.

    - Translation updates
    Hungarian, Bulgarian, Polish, Slovak, French, Chinese (Taiwan), Galician,
    Italian

  • List of bugs fixed since Clutter 1.25.4

    #752752 - Patches needed for relative pointer and pointer confinement
              in mutter
    #755849 - evdev: Recycle device ids after devices are removed

Many thanks to:

  Jonas Ådahl, Alexander Shopov, Alexandre Franke, Chao-Hsiung Liao,
  Dušan Kazik, Fran Dieguez, Gábor Kelemen, Milo Casagrande, Piotr Drąg

Clutter 1.25.4                                                       2016-01-21
===============================================================================

  • List of changes since Clutter 1.25.2

    - Add more scrolling information inside ClutterScrollEvent
    This allows to receive the type of device that generate the scroll event,
    as well as the notification that a scroll motion finished; we can use this
    information to implement kinetic scrolling. Currently, only libinput
    provides us with these fields.

    - Translation updates
    Spanish, Lithuanian, Czech, German, Hebrew, Brazilian Portuguese

  • List of bugs fixed since Clutter 1.25.2

     #757026 - Add scroll source / axis stop information to scroll events

Many thanks to:

  Aurimas Černius, Carlos Garnacho, Daniel Mustieles, Marek Černocký,
  Mario Blättermann, Rafael Fontenelle, Yosef Or Boczko

Clutter 1.25.2                                                       2015-12-17
===============================================================================

  • List of changes since Clutter 1.24

    - Set event code on button and key events in the evdev input backend
    This allows users of the evdev input backend to retrieve the event
    code without doing ad hoc calculations based on event-specific fields.

    - Take scaling factor into account on Wayland with GDK
    This brings the Wayland support inside the GDK backend on par to the X11
    side.

    - Decouple selectability from editability in ClutterText
    It's possible to create selectable ClutterText actors without enabling
    editability; this may cause unexpected cursors to appear in reactive
    label-like actors using non-editable ClutterText, so feedback from users
    is appreciated.

    - Allow overriding the Cogl driver used by Clutter
    On systems with multiple OpenGL and OpenGL ES implementations available,
    it should be possible to explicitly set the Cogl driver (OpenGL 3.2 core

clutter/clutter-version.h
/*
 * Clutter.
 *
 * An OpenGL based 'interactive canvas' library.
 *
 * Authored By Matthew Allum  <mallum@openedhand.com>
 *
 * Copyright (C) 2006 OpenedHand
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library. If not, see <http://www.gnu.org/licenses/>.
 *
 *
 */

#ifndef __CLUTTER_VERSION_H__
#define __CLUTTER_VERSION_H__

#if !defined(__CLUTTER_H_INSIDE__) && !defined(CLUTTER_COMPILATION)
#error "Only <clutter/clutter.h> can be included directly."
#endif

/**
 * SECTION:clutter-version
 * @short_description: Versioning utility macros
 *
 * Clutter offers a set of macros for checking the version of the library
 * at compile time; it also provides a function to perform the same check
 * at run time.
 *
 * Clutter adds version information to both API deprecations and additions;
 * by definining the macros %CLUTTER_VERSION_MIN_REQUIRED and
 * %CLUTTER_VERSION_MAX_ALLOWED, you can specify the range of Clutter versions
 * whose API you want to use. Functions that were deprecated before, or
 * introduced after, this range will trigger compiler warnings. For instance,
 * if we define the following symbols:
 *
 * |[
 *   CLUTTER_VERSION_MIN_REQUIRED = CLUTTER_VERSION_1_6
 *   CLUTTER_VERSION_MAX_ALLOWED  = CLUTTER_VERSION_1_8
 * ]|
 *
 * and we have the following functions annotated in the Clutter headers:
 *
 * |[
 *   void clutter_function_A (void) CLUTTER_DEPRECATED_IN_1_4;
 *   void clutter_function_B (void) CLUTTER_DEPRECATED_IN_1_6;
 *   void clutter_function_C (void) CLUTTER_AVAILABLE_IN_1_8;
 *   void clutter_function_D (void) CLUTTER_AVAILABLE_IN_1_10;
 * ]|
 *
 * then any application code using the functions above will get the output:
 *
 * |[
 *   clutter_function_A: deprecation warning
 *   clutter_function_B: no warning
 *   clutter_function_C: no warning
 *   clutter_function_D: symbol not available warning
 * ]|
 *
 * It is possible to disable the compiler warnings by defining the macro
 * %CLUTTER_DISABLE_DEPRECATION_WARNINGS before including the clutter.h
 * header.
 */

#include <glib.h>

G_BEGIN_DECLS

/**
 * CLUTTER_MAJOR_VERSION:
 *
 * The major version of the Clutter library (1, if %CLUTTER_VERSION is 1.2.3)
 */
#define CLUTTER_MAJOR_VERSION   (1)

/**
 * CLUTTER_MINOR_VERSION:
 *
 * The minor version of the Clutter library (2, if %CLUTTER_VERSION is 1.2.3)
 */
#define CLUTTER_MINOR_VERSION   (26)

/**
 * CLUTTER_MICRO_VERSION:
 *
 * The micro version of the Clutter library (3, if %CLUTTER_VERSION is 1.2.3)
 */
#define CLUTTER_MICRO_VERSION   (2)

/**
 * CLUTTER_VERSION:
 *
 * The full version of the Clutter library, like 1.2.3
 */
#define CLUTTER_VERSION         1.26.2

/**
 * CLUTTER_VERSION_S:
 *
 * The full version of the Clutter library, in string form (suited for
 * string concatenation)
 */
#define CLUTTER_VERSION_S       "1.26.2"

/**
 * CLUTTER_VERSION_HEX:
 *
 * Numerically encoded version of the Clutter library, like 0x010203
 */
#define CLUTTER_VERSION_HEX     ((CLUTTER_MAJOR_VERSION << 24) | \
                                 (CLUTTER_MINOR_VERSION << 16) | \
                                 (CLUTTER_MICRO_VERSION << 8))

/* XXX - Every new stable minor release bump should add a macro here */

/**
 * CLUTTER_VERSION_1_0:
 *
 * A macro that evaluates to the 1.0 version of Clutter, in a format
 * that can be used by the C pre-processor.
 *
 * Since: 1.10
 */
#define CLUTTER_VERSION_1_0     (G_ENCODE_VERSION (1, 0))

/**
 * CLUTTER_VERSION_1_2:
 *
 * A macro that evaluates to the 1.2 version of Clutter, in a format
 * that can be used by the C pre-processor.
 *
 * Since: 1.10
 */
#define CLUTTER_VERSION_1_2     (G_ENCODE_VERSION (1, 2))

/**
 * CLUTTER_VERSION_1_4:
 *
 * A macro that evaluates to the 1.4 version of Clutter, in a format
 * that can be used by the C pre-processor.
 *
 * Since: 1.10
 */
#define CLUTTER_VERSION_1_4     (G_ENCODE_VERSION (1, 4))

/**
 * CLUTTER_VERSION_1_6:
 *
 * A macro that evaluates to the 1.6 version of Clutter, in a format
 * that can be used by the C pre-processor.
 *
 * Since: 1.10
 */
#define CLUTTER_VERSION_1_6     (G_ENCODE_VERSION (1, 6))

/**
 * CLUTTER_VERSION_1_8:
 *
 * A macro that evaluates to the 1.8 version of Clutter, in a format
 * that can be used by the C pre-processor.
 *
 * Since: 1.10
 */
#define CLUTTER_VERSION_1_8     (G_ENCODE_VERSION (1, 8))

/**
 * CLUTTER_VERSION_1_10:
 *
 * A macro that evaluates to the 1.10 version of Clutter, in a format
 * that can be used by the C pre-processor.
 *
 * Since: 1.10
 */
#define CLUTTER_VERSION_1_10    (G_ENCODE_VERSION (1, 10))

/**
 * CLUTTER_VERSION_1_12:
 *
 * A macro that evaluates to the 1.12 version of Clutter, in a format
 * that can be used by the C pre-processor.
 *
 * Since: 1.12
 */
#define CLUTTER_VERSION_1_12    (G_ENCODE_VERSION (1, 12))

/**
 * CLUTTER_VERSION_1_14:
 *
 * A macro that evaluates to the 1.14 version of Clutter, in a format
 * that can be used by the C pre-processor.
 *
 * Since: 1.14
 */
#define CLUTTER_VERSION_1_14    (G_ENCODE_VERSION (1, 14))

/**
 * CLUTTER_VERSION_1_16:
 *
 * A macro that evaluates to the 1.16 version of Clutter, in a format
 * that can be used by the C pre-processor.
 *
 * Since: 1.16
 */
#define CLUTTER_VERSION_1_16    (G_ENCODE_VERSION (1, 16))

/**
 * CLUTTER_VERSION_1_18:
 *
 * A macro that evaluates to the 1.18 version of Clutter, in a format
 * that can be used by the C pre-processor.
 *
 * Since: 1.18
 */
#define CLUTTER_VERSION_1_18    (G_ENCODE_VERSION (1, 18))

/**
 * CLUTTER_VERSION_1_20:
 *
 * A macro that evaluates to the 1.20 version of Clutter, in a format
 * that can be used by the C pre-processor.
 *
 * Since: 1.20
 */
#define CLUTTER_VERSION_1_20    (G_ENCODE_VERSION (1, 20))

/**
 * CLUTTER_VERSION_1_22:
 *
 * A macro that evaluates to the 1.22 version of Clutter, in a format
 * that can be used by the C pre-processor.
 *
 * Since: 1.22
 */
#define CLUTTER_VERSION_1_22    (G_ENCODE_VERSION (1, 22))

/**
 * CLUTTER_VERSION_1_24:
 *
 * A macro that evaluates to the 1.24 version of Clutter, in a format
 * that can be used by the C pre-processor.
 *
 * Since: 1.24
 */
#define CLUTTER_VERSION_1_24    (G_ENCODE_VERSION (1, 24))

/**
 * CLUTTER_VERSION_1_26:
 *
 * A macro that evaluates to the 1.26 version of Clutter, in a format
 * that can be used by the C pre-processor.
 *
 * Since: 1.26
 */
#define CLUTTER_VERSION_1_26    (G_ENCODE_VERSION (1, 26))

/* evaluates to the current stable version; for development cycles,
 * this means the next stable target
 */
#if (CLUTTER_MINOR_VERSION % 2)
# define CLUTTER_VERSION_CUR_STABLE      (G_ENCODE_VERSION (CLUTTER_MAJOR_VERSION, CLUTTER_MINOR_VERSION + 1))
#else
# define CLUTTER_VERSION_CUR_STABLE      (G_ENCODE_VERSION (CLUTTER_MAJOR_VERSION, CLUTTER_MINOR_VERSION))
#endif

/* evaluates to the previous stable version */
#if (CLUTTER_MINOR_VERSION % 2)
# define CLUTTER_VERSION_PREV_STABLE     (G_ENCODE_VERSION (CLUTTER_MAJOR_VERSION, CLUTTER_MINOR_VERSION - 1))
#else
# define CLUTTER_VERSION_PREV_STABLE     (G_ENCODE_VERSION (CLUTTER_MAJOR_VERSION, CLUTTER_MINOR_VERSION - 2))
#endif

/**
 * CLUTTER_CHECK_VERSION:
 * @major: major version, like 1 in 1.2.3
 * @minor: minor version, like 2 in 1.2.3
 * @micro: micro version, like 3 in 1.2.3
 *
 * Evaluates to %TRUE if the version of the Clutter library is greater
 * than @major, @minor and @micro
 */
#define CLUTTER_CHECK_VERSION(major,minor,micro) \
        (CLUTTER_MAJOR_VERSION > (major) || \
         (CLUTTER_MAJOR_VERSION == (major) && CLUTTER_MINOR_VERSION > (minor)) || \
         (CLUTTER_MAJOR_VERSION == (major) && CLUTTER_MINOR_VERSION == (minor) && CLUTTER_MICRO_VERSION >= (micro)))

#ifndef _CLUTTER_EXTERN
#define _CLUTTER_EXTERN extern
#endif

#ifdef CLUTTER_WINDOWING_WIN32
# ifdef CLUTTER_COMPILATION
#  ifdef DLL_EXPORT
#   define CLUTTER_VAR __declspec(dllexport)
#  else
#   define CLUTTER_VAR extern
#  endif
# else
#  define CLUTTER_VAR __declspec(dllimport)
# endif
#else
# define CLUTTER_VAR _CLUTTER_EXTERN
#endif

/**
 * clutter_major_version:
 *
 * The major component of the Clutter library version, e.g. 1 if the version
 * is 1.2.3
 *
 * This value can be used for run-time version checks
 *
 * For a compile-time check, use %CLUTTER_MAJOR_VERSION
 *
 * Since: 1.2
 */
CLUTTER_VAR const guint clutter_major_version;

/**
 * clutter_minor_version:
 *
 * The minor component of the Clutter library version, e.g. 2 if the version
 * is 1.2.3
 *
 * This value can be used for run-time version checks
 *
 * For a compile-time check, use %CLUTTER_MINOR_VERSION
 *
 * Since: 1.2
 */
CLUTTER_VAR const guint clutter_minor_version;

/**
 * clutter_micro_version:
 *
 * The micro component of the Clutter library version, e.g. 3 if the version
 * is 1.2.3
 *
 * This value can be used for run-time version checks
 *
 * For a compile-time check, use %CLUTTER_MICRO_VERSION
 *
 * Since: 1.2
 */
CLUTTER_VAR const guint clutter_micro_version;

G_END_DECLS

#endif /* __CLUTTER_VERSION_H__ */
